File: app/src/main/java/com/martinlaizg/geofind/data/access/database/entities/Place.java
package com.martinlaizg.geofind.data.access.database.entities;

import androidx.room.Entity;
import androidx.room.ForeignKey;
import androidx.room.Ignore;
import androidx.room.Index;
import androidx.room.PrimaryKey;

import com.google.android.gms.maps.model.LatLng;
import com.martinlaizg.geofind.utils.DateUtils;

import java.sql.Date;
import java.util.Calendar;

@Entity(tableName = "places", foreignKeys = @ForeignKey(entity = Tour.class, parentColumns = "id",
                                                        childColumns = "tour_id"),
        indices = @Index("tour_id"))
public class Place {

	@PrimaryKey
	private final Integer id;
	private String name;
	private Double lat;
	private Double lon;
	private Integer tour_id;
	private String description;
	private Integer order;
	private Date created_at;
	private Date updated_at;
	private Date updated;

	public Place(Integer id, String name, Double lat, Double lon, Integer tour_id,
			String description, Integer order, Date created_at, Date updated_at, Date updated) {
		this.id = id;
		this.name = name;
		this.lat = lat;
		this.lon = lon;
		this.tour_id = tour_id;
		this.description = description;
		this.order = order;
		this.created_at = created_at;
		this.updated_at = updated_at;
		this.updated = updated;
	}

	@Ignore
	public Place() {
		id = 0;
	}

	public Integer getId() {
		return id;
	}

	public Date getCreated_at() {
		return created_at;
	}

	public void setCreated_at(Date created_at) {
		this.created_at = created_at;
	}

	public Date getUpdated_at() {
		return updated_at;
	}

	public void setUpdated_at(Date updated_at) {
		this.updated_at = updated_at;
	}

	public Integer getTour_id() {
		return tour_id;
	}

	public void setTour_id(Integer tour_id) {
		this.tour_id = tour_id;
	}

	public LatLng getPosition() {
		return new LatLng(getLat(), getLon());
	}

	public Double getLat() {
		return lat;
	}

	public Double getLon() {
		return lon;
	}

	public void setPosition(LatLng position) {
		setLat(position.latitude);
		setLon(position.longitude);
	}

	public void setLat(Double lat) {
		this.lat = lat;
	}

	public void setLon(Double lon) {
		this.lon = lon;
	}

	public Integer getOrder() {
		return order;
	}

	public void setOrder(Integer order) {
		this.order = order;
	}

	boolean isValid() {
		return getName() != null && !getName().isEmpty() && getDescription() != null &&
				!getDescription().isEmpty() && getLat() != null && getLon() != null;
	}

	public String getName() {
		return name;
	}

	public void setName(String name) {
		this.name = name;
	}

	public String getDescription() {
		return description;
	}

	public void setDescription(String description) {
		this.description = description;
	}

	public Date getUpdated() {
		return new Date(Calendar.getInstance().getTime().getTime());
	}

	public void setUpdated(Date updated) {
		if(updated == null) {
			this.updated = new Date(Calendar.getInstance().getTime().getTime());
		}
		this.updated = updated;
	}

	public boolean isOutOfDate() {
		return DateUtils.isDateExpire(updated);
	}

}

File: app/src/main/java/com/martinlaizg/geofind/data/access/database/entities/Place.kt
package com.martinlaizg.geofind.data.access.database.entities

import android.location.Location
import androidx.room.*
import com.google.android.gms.maps.model.LatLng

@Entity(tableName = "places",
		foreignKeys = [ForeignKey(entity = Tour::class, parentColumns = ["id"], childColumns = ["tour_id"], onDelete = ForeignKey.CASCADE)],
		indices = [Index("tour_id")])
class Place {
	@PrimaryKey
	val id: Int
	var name: String
	var description: String

	private var lat: Double? = null
	private var lon: Double? = null
	private var tourId: Int? = null
	var order: Int? = null

	// Question
	var question: String? = null
	private var answer: String? = null
	private var answer2: String? = null
	private var answer3: String? = null
	private var image: String? = null

	constructor(id: Int, name: String, description: String) {
		this.id = id
		this.name = name
		this.description = description
	}

	@Ignore
	constructor() {
		id = 0
		name = ""
		description = ""
	}

	var position: LatLng?
		get() {
			return if (lat == null || lon == null) {
				null
			} else {
				LatLng(lat!!, lon!!)
			}
		}
		set(value) {
			lat = value?.latitude
			lon = value?.longitude
		}

	val isValid: Boolean
		get() = name.isNotEmpty() && description.isNotEmpty() && lat != null && lon != null

	var location: Location?
		get() {
			if (lat == null || lon == null) return null
			val l = Location("")
			l.latitude = lat!!
			l.longitude = lon!!
			return l
		}
		set(location) {
			if (location == null) return
			lat = location.latitude
			lon = location.longitude
		}

	fun fillQuestion(newQuestion: String = "", correctAnswer: String = "", secondAnswer: String = "", thirdAnswer: String = "") {
		question = newQuestion
		answer = correctAnswer
		answer2 = secondAnswer
		answer3 = thirdAnswer
	}
}